### PR TITLE
sh - link to working form with validation

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by seanhelvey**
This commit covers:
- An images resource (route, controller, views, and model, along with schema and migration). 
- Model, controller, and routes have decent coverage. Routes tests are in integration folder.
- If anyone is up for pairing on a couple of high level tests with Capybara, that would be awesome.

Closes #2 

img

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   root 'application#home'
+
+  resources :images
```
[the remark this answers is not in the document]
Comment (seanhelvey) [reply config/routes.rb]: derp! Fixed.

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    @image = Image.new(image_params)
+    if @image.valid?
+      @image.save
+      redirect_to @image
+    else
+      render 'new'
+    end
+  end
+
+  def image_params
```
[the remark this answers is not in the document]
Comment (seanhelvey) [reply app/controllers/images_controller.rb]: Yup thank you

File: app/assets/javascripts/images.js [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
[the remark this answers is not in the document]
Comment (seanhelvey) [reply app/assets/javascripts/images.js]: removed

File: app/views/home/index.html.erb [hunk 4]
```diff
@@ -0,0 +1 @@
+<a href="images/new">Add an image<a>
```
[the remark this answers is not in the document]
Comment (seanhelvey) [reply app/views/home/index.html.erb]: done

File: db/migrate/20190402210500_create_images.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class CreateImages < ActiveRecord::Migration[5.2]
+  def change
+    create_table :images do |t|
+      t.string :url
```
[the remark this answers is not in the document]
Comment (seanhelvey) [reply db/migrate/20190402210500_create_images.rb]: Fixed and tested

**Comment by seanhelvey** [comment]
Fixed everything you pointed out @grahamotte thank you!

File: app/assets/stylesheets/images.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Images controller here.
```
Comment (grahamotte) [inline app/assets/stylesheets/images.scss]: Another empty file

File: app/models/image.rb [hunk 7]
```diff
@@ -0,0 +1,3 @@
+class Image < ApplicationRecord
+  validates :url, presence: true, null: false, format: { with: URI.regexp(%w[http https]) }
```
Comment (seanhelvey) [inline app/models/image.rb]: @grahamotte isn't this what you suggested ^^

File: db/migrate/20190402210500_create_images.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (seanhelvey) [reply db/migrate/20190402210500_create_images.rb]: Added to migration also ✔️

File: test/models/image_test.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  test 'image is valid' do
```
Comment (joelbandi) [inline test/models/image_test.rb]: We generally use the `def test_image_is_valid`  format to write tests

File: app/views/images/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+<%= form_with scope: :image, url: images_path, local: true do |form| %>
```
Comment (joelbandi) [inline app/views/images/new.html.erb]: `simple_form`` is great for forms that mirror model attributes!

File: test/models/image_test.rb (hunk 8 above)
Comment (seanhelvey) [reply test/models/image_test.rb]: Thanks! Fixed.

File: app/views/images/new.html.erb (hunk 9 above)
Comment (seanhelvey) [reply app/views/images/new.html.erb]: Done!

**Comment by seanhelvey** [comment]
@joelbandi thanks for the feedback! Should be 👍

**Pull request merged**